mingw64 8.1 ambiguous string_view construction

Unable to build boost.url under mingw64 8.1
Boost: 1.81
compiler:
```
C:/Qt/Tools/mingw810_32/bin/gcc.exe --version
gcc.exe (i686-posix-dwarf-rev0, Built by MinGW-W64 project) 8.1.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
error message:
```
In file included from $BUILD_PATH/boost-src/libs/url/include/boost/url.hpp:16,
                 from $BUILD_PATH/boost-src/libs/url/include/boost/url/src.hpp:28,
                 from $BUILD_PATH/boost-src/libs/url/src/src.cpp:10:
$BUILD_PATH/boost-src/libs/url/include/boost/url/decode_view.hpp: In constructor 'boost::urls::decode_view::decode_view(boost::urls::pct_string_view, boost::urls::encoding_opts)':
$BUILD_PATH/boost-src/libs/url/include/boost/url/decode_view.hpp:215:26: error: call of overloaded 'basic_string_view(boost::urls::pct_string_view&)' is ambiguous
             string_view(s),
                          ^
In file included from $BUILD_PATH/boost-src/libs/url/include/boost/url/string_view.hpp:14,
                 from $BUILD_PATH/boost-src/libs/url/include/boost/url/grammar/ci_string.hpp:15,
                 from $BUILD_PATH/boost-src/libs/url/include/boost/url/grammar.hpp:18,
                 from $BUILD_PATH/boost-src/libs/url/include/boost/url.hpp:13,
                 from $BUILD_PATH/boost-src/libs/url/include/boost/url/src.hpp:28,
                 from $BUILD_PATH/boost-src/libs/url/src/src.cpp:10:
$BUILD_PATH/boost-src/libs/core/include/boost/core/detail/string_view.hpp:395:5: note: candidate: 'boost::core::basic_string_view<Ch>::basic_string_view(const std::basic_string_view<Ch, std::char_traits<_CharT> >&) [with Ch = char]'
     basic_string_view( std::basic_string_view<Ch, std::char_traits<Ch> > const& str ) BOOST_NOEXCEPT: p_( str.data() ), n_( str.size() )
     ^~~~~~~~~~~~~~~~~
$BUILD_PATH/boost-src/libs/core/include/boost/core/detail/string_view.hpp:339:26: note: candidate: 'constexpr boost::core::basic_string_view<char>::basic_string_view(const boost::core::basic_string_view<char>&)'
 template<class Ch> class basic_string_view
                          ^~~~~~~~~~~~~~~~~
$BUILD_PATH/boost-src/libs/core/include/boost/core/detail/string_view.hpp:339:26: note: candidate: 'constexpr boost::core::basic_string_view<char>::basic_string_view(boost::core::basic_string_view<char>&&)'
In file included from $BUILD_PATH/boost-src/libs/url/include/boost/url/src.hpp:87,
                 from $BUILD_PATH/boost-src/libs/url/src/src.cpp:10:
$BUILD_PATH/boost-src/libs/url/include/boost/url/grammar/impl/string_view_base.ipp: In function 'std::ostream& boost::urls::grammar::operator<<(std::ostream&, const boost::urls::grammar::string_view_base&)':
$BUILD_PATH/boost-src/libs/url/include/boost/url/grammar/impl/string_view_base.ipp:25:31: error: call of overloaded 'basic_string_view(const boost::urls::grammar::string_view_base&)' is ambiguous
     return os << string_view(s);
                               ^
```

--- a/.drone.star
+++ b/.drone.star
@@ -13,17 +13,19 @@
 def main(ctx):
     return generate(
         # Compilers
-        ['gcc >=4.8',
-         'clang >=3.8',
-         'msvc >=14.1',
-         'arm64-gcc latest',
-         's390x-gcc latest',
-         # 'freebsd-gcc latest',
-         'apple-clang *',
-         'arm64-clang latest',
-         's390x-clang latest',
-         'freebsd-clang latest',
-         'x86-msvc latest'],
+        [
+            'gcc >=4.8',
+            'clang >=3.8',
+            'msvc >=14.1',
+            'arm64-gcc latest',
+            's390x-gcc latest',
+            # 'freebsd-gcc latest',
+            'apple-clang *',
+            'arm64-clang latest',
+            's390x-clang latest',
+            'freebsd-clang latest',
+            'x86-msvc latest'
+        ],
         # Standards
         '>=11', cache_dir='cache') + [
                linux_cxx("GCC 12 (no-mutex)", "g++-12", packages="g++-12", buildscript="drone", buildtype="boost",
@@ -52,8 +54,8 @@ load("@boost_ci//ci/drone/:functions.star", "linux_cxx", "windows_cxx", "osx_cxx
 def generate(compiler_ranges, cxx_range, max_cxx=2, coverage=True, docs=True, asan=True, tsan=False, ubsan=True,
              cmake=True, cache_dir=None, packages=[]):
     # swap: `packages` variable will be reused
-    packages_requested=list(packages)
-    packages=[]
+    packages_requested = list(packages)
+    packages = []
 
     # Get compiler versions we should test
     # Each compiler is a tuple (represented as a list) of:
@@ -183,7 +185,15 @@ def generate(compiler_ranges, cxx_range, max_cxx=2, coverage=True, docs=True, as
             packages = ['docbook', 'docbook-xml', 'docbook-xsl', 'xsltproc', 'libsaxonhe-java', 'default-jre-headless',
                         'flex', 'libfl-dev', 'bison', 'unzip', 'rsync']
         elif compiler.endswith('gcc'):
-            if version[0] >= 5:
+            if version_str == '8.0.1':
+                # The image supports 8.0.1 and 8.4 via apt-get
+                # Explicitly testing 8.0.1 is useful since gcc >=8 <8.4 contains a bug regarding ambiguity in implicit
+                # conversions to boost::core::string_view in C++17 mode.
+                packages = ['--allow-downgrades', 'libmpx2=8-20180414-1ubuntu2', 'libgcc-8-dev=8-20180414-1ubuntu2',
+                            'cpp-8=8-20180414-1ubuntu2',
+                            'gcc-8-base=8-20180414-1ubuntu2', 'gcc-8=8-20180414-1ubuntu2', 'g++-8=8-20180414-1ubuntu2',
+                            'libstdc++-8-dev=8-20180414-1ubuntu2']
+            elif version[0] >= 5:
                 packages = ['g++-' + str(version[0])]
             else:
                 packages = ['g++-' + str(version[0]) + '.' + str(version[1])]
@@ -506,7 +516,7 @@ def image_supports_caching(image_str, compiler_str):
 # - compilers_in_range('gcc >=10') -> [('gcc', '12'), ('gcc', '11'), ('gcc', '10')]
 def compilers_in_range(compiler_range_str):
     supported = {
-        'gcc': ['12', '11', '10', '9', '8', '7', '6', '5', '4.9', '4.8'],
+        'gcc': ['12', '11', '10', '9', '8.4', '8.0.1', '7', '6', '5', '4.9', '4.8'],
         's390x-gcc': ['12'],
         'arm64-gcc': ['11'],
         'freebsd-gcc': ['11', '10', '9', '8'],

--- a/.drone/drone.sh
+++ b/.drone/drone.sh
@@ -6,6 +6,8 @@
 
 set -xe
 
+echo "==================================> ENVIRONMENT"
+
 TRAVIS_BUILD_DIR=$(pwd)
 export TRAVIS_BUILD_DIR
 DRONE_BUILD_DIR=$(pwd)
@@ -20,6 +22,21 @@ export USER
 export CC=${CC:-gcc}
 export PATH=~/.local/bin:/usr/local/bin:$PATH
 printenv
+
+echo "==================================> PACKAGES"
+
+for package in ${PACKAGES// / }; do
+  if [[ "$package" == "--"* ]]; then
+    continue
+  fi
+  package_no_version=${package%%=*}
+  echo "Versions available for $package_no_version"
+  apt-cache policy "$package_no_version" || true
+done
+
+if command -v "$CXX" &>/dev/null; then
+  $CXX --version
+fi
 
 common_install() {
   if [ "$TRAVIS_OS_NAME" == "osx" ]; then

--- a/example/finicky/finicky.cpp
+++ b/example/finicky/finicky.cpp
@@ -251,7 +251,7 @@ int main(int argc, char** argv)
                         it->value().is_string(),
                         "authority field should be a string");
                     u.set_encoded_authority(
-                        it->value().as_string());
+                        it->value().as_string().subview());
                 }
 
                 it = uo.find("username");
@@ -263,7 +263,7 @@ int main(int argc, char** argv)
                         it->value().is_string(),
                         "username field should be a string");
                     u.set_encoded_user(
-                        it->value().as_string());
+                        it->value().as_string().subview());
                 }
 
                 it = uo.find("password");
@@ -273,7 +273,7 @@ int main(int argc, char** argv)
                         it->value().is_string(),
                         "password field should be a string");
                     u.set_encoded_password(
-                        it->value().as_string());
+                        it->value().as_string().subview());
                 }
 
                 it = uo.find("userinfo");
@@ -283,7 +283,7 @@ int main(int argc, char** argv)
                         it->value().is_string(),
                         "userinfo field should be a string");
                     u.set_encoded_userinfo(
-                        it->value().as_string());
+                        it->value().as_string().subview());
                 }
 
                 it = uo.find("host");
@@ -293,7 +293,7 @@ int main(int argc, char** argv)
                         it->value().is_string(),
                         "host field should be a string");
                     u.set_encoded_host(
-                        it->value().as_string());
+                        it->value().as_string().subview());
                 }
 
                 it = uo.find("port");
@@ -303,7 +303,7 @@ int main(int argc, char** argv)
                         it->value().is_string(),
                         "port field should be a string");
                     u.set_port(
-                        it->value().as_string());
+                        it->value().as_string().subview());
                 }
 
                 it = uo.find("path");
@@ -315,7 +315,7 @@ int main(int argc, char** argv)
                         it->value().is_string(),
                         "path field should be a string");
                     u.set_encoded_path(
-                        it->value().as_string());
+                        it->value().as_string().subview());
                 }
 
                 it = uo.find("query");
@@ -327,7 +327,7 @@ int main(int argc, char** argv)
                         it->value().is_string(),
                         "query field should be a string");
                     u.set_encoded_query(
-                        it->value().as_string());
+                        it->value().as_string().subview());
                 }
 
                 it = uo.find("fragment");
@@ -339,7 +339,7 @@ int main(int argc, char** argv)
                         it->value().is_string(),
                         "fragment field should be a string");
                     u.set_encoded_fragment(
-                        it->value().as_string());
+                        it->value().as_string().subview());
                 }
             }
         }
@@ -379,7 +379,7 @@ int main(int argc, char** argv)
             // Look for match and change browser
             if (url_match(mit->value(), u))
             {
-                browser = hbit->value().as_string();
+                browser = hbit->value().as_string().subview();
                 break;
             }
         }

--- a/example/mailto/rfc.hpp
+++ b/example/mailto/rfc.hpp
@@ -356,7 +356,7 @@ struct hfname_rule_t
         it += s.size();
         auto r = grammar::parse(s, hfvalue_rule);
         if (!r)
-            return r;
+            return r.error();
 
         // The user agent interpreting a 'mailto' URI SHOULD NOT create a
         // message if any of the header fields are considered dangerous

--- a/include/boost/url/decode_view.hpp
+++ b/include/boost/url/decode_view.hpp
@@ -211,7 +211,7 @@ public:
         pct_string_view s,
         encoding_opts opt = {}) noexcept
         : decode_view(
-            string_view(s),
+            detail::to_sv(s),
             s.decoded_size(),
             opt)
     {

--- a/include/boost/url/detail/any_segments_iter.hpp
+++ b/include/boost/url/detail/any_segments_iter.hpp
@@ -207,7 +207,7 @@ private:
         if(it_ == end_)
             return false;
         measure_impl(n,
-            string_view(*it_),
+            detail::to_sv(*it_),
             encode_colons);
         ++it_;
         return true;
@@ -219,7 +219,7 @@ private:
         char const* end) noexcept override
     {
         copy_impl(dest, end,
-            string_view(*it_++),
+            detail::to_sv(*it_++),
             encode_colons);
     }
 };
@@ -292,7 +292,7 @@ struct segments_encoded_iter
         {
             // throw on invalid input
             front = pct_string_view(
-                string_view(*first));
+                detail::to_sv(*first));
             auto it = first;
             if(++it == last)
                 fast_nseg = 1;
@@ -325,7 +325,7 @@ private:
         // throw on invalid input
         measure_impl(n,
             pct_string_view(
-                string_view(*it_++)),
+                detail::to_sv(*it_++)),
             encode_colons);
         return true;
     }
@@ -336,7 +336,7 @@ private:
         char const* end) noexcept override
     {
         copy_impl(dest, end,
-            string_view(*it_++),
+            detail::to_sv(*it_++),
             encode_colons);
     }
 };

--- a/include/boost/url/detail/impl/pattern.ipp
+++ b/include/boost/url/detail/impl/pattern.ipp
@@ -592,8 +592,7 @@ struct host_template_rule_t
             // any_host_template_rule can always
             // be empty, so it's never invalid
             BOOST_ASSERT(rv);
-            return *rv;
-
+            return detail::to_sv(*rv);
         }
         // IP-literals need to be enclosed in
         // "[]" if using ':' in the template
@@ -619,7 +618,6 @@ struct host_template_rule_t
         // closing "]"
         BOOST_ASSERT(rv);
         return string_view{it0, it};
-
     }
 };
 

--- a/include/boost/url/detail/optional_string.hpp
+++ b/include/boost/url/detail/optional_string.hpp
@@ -69,7 +69,7 @@ get_optional_string(
         std::is_convertible<typename std::decay<decltype(*std::declval<OptionalString>())>::type, string_view>::value,
         "OptionalString requirements not met");
     optional_string r;
-    r.s = opt ? string_view(*opt) : string_view{};
+    r.s = opt ? detail::to_sv(*opt) : string_view{};
     r.b = static_cast<bool>(opt);
     return r;
 }

--- a/include/boost/url/detail/string_view.hpp
+++ b/include/boost/url/detail/string_view.hpp
@@ -1,0 +1,34 @@
+//
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/url
+//
+
+#ifndef BOOST_URL_DETAIL_STRING_VIEW_HPP
+#define BOOST_URL_DETAIL_STRING_VIEW_HPP
+
+#include <boost/core/detail/string_view.hpp>
+
+namespace boost {
+namespace urls {
+namespace detail {
+
+// We use detail::to_sv(s) instead of string_view(s) whenever
+// we should convert to string_view.
+// This is a workaround for GCC >=8.0 <8.4
+// See: https://github.com/boostorg/url/issues/672
+template<class T>
+core::string_view
+to_sv(T const& t) noexcept
+{
+    return boost::core::string_view(t);
+}
+
+} // detail
+} // urls
+} // boost
+
+#endif

--- a/include/boost/url/grammar/impl/string_view_base.ipp
+++ b/include/boost/url/grammar/impl/string_view_base.ipp
@@ -22,7 +22,7 @@ operator<<(
     std::ostream& os,
     string_view_base const& s)
 {
-    return os << string_view(s);
+    return os << urls::detail::to_sv(s);
 }
 
 } // grammar

--- a/include/boost/url/grammar/string_view_base.hpp
+++ b/include/boost/url/grammar/string_view_base.hpp
@@ -777,7 +777,7 @@ public:
         typename std::enable_if<
             is_match<S0, S1>::value, bool>::type
     {
-        return string_view(s0) == string_view(s1);
+        return urls::detail::to_sv(s0) == urls::detail::to_sv(s1);
     }
 
     template<class S0, class S1>
@@ -786,7 +786,7 @@ public:
         typename std::enable_if<
             is_match<S0, S1>::value, bool>::type
     {
-        return string_view(s0) != string_view(s1);
+        return urls::detail::to_sv(s0) != urls::detail::to_sv(s1);
     }
 
     template<class S0, class S1>
@@ -795,7 +795,7 @@ public:
         typename std::enable_if<
             is_match<S0, S1>::value, bool>::type
     {
-        return string_view(s0) < string_view(s1);
+        return urls::detail::to_sv(s0) < urls::detail::to_sv(s1);
     }
 
     template<class S0, class S1>
@@ -804,7 +804,7 @@ public:
         typename std::enable_if<
             is_match<S0, S1>::value, bool>::type
     {
-        return string_view(s0) <= string_view(s1);
+        return urls::detail::to_sv(s0) <= urls::detail::to_sv(s1);
     }
 
     template<class S0, class S1>
@@ -813,7 +813,7 @@ public:
         typename std::enable_if<
             is_match<S0, S1>::value, bool>::type
     {
-        return string_view(s0) > string_view(s1);
+        return urls::detail::to_sv(s0) > urls::detail::to_sv(s1);
     }
 
     template<class S0, class S1>
@@ -822,7 +822,7 @@ public:
         typename std::enable_if<
             is_match<S0, S1>::value, bool>::type
     {
-        return string_view(s0) >= string_view(s1);
+        return urls::detail::to_sv(s0) >= urls::detail::to_sv(s1);
     }
 #endif
 
@@ -857,6 +857,19 @@ operator<<(
     string_view_base const& s);
 
 } // grammar
+
+#ifndef BOOST_URL_DOCS
+namespace detail {
+template <>
+inline
+string_view
+to_sv(grammar::string_view_base const& s) noexcept
+{
+    return s.operator string_view();
+}
+} // detail
+#endif
+
 } // urls
 } // boost
 

--- a/include/boost/url/pct_string_view.hpp
+++ b/include/boost/url/pct_string_view.hpp
@@ -176,7 +176,7 @@ public:
     pct_string_view(
         String const& s)
         : pct_string_view(
-            string_view(s))
+            detail::to_sv(s))
     {
     }
 
@@ -442,6 +442,18 @@ make_pct_string_view_unsafe(
     return pct_string_view(
         data, size, decoded_size);
 }
+#endif
+
+#ifndef BOOST_URL_DOCS
+namespace detail {
+template <>
+inline
+string_view
+to_sv(pct_string_view const& s) noexcept
+{
+    return s.substr();
+}
+} // detail
 #endif
 
 } // urls

--- a/include/boost/url/string_view.hpp
+++ b/include/boost/url/string_view.hpp
@@ -11,7 +11,7 @@
 #define BOOST_URL_STRING_VIEW_HPP
 
 #include <boost/url/detail/config.hpp>
-#include <boost/core/detail/string_view.hpp>
+#include <boost/url/detail/string_view.hpp>
 
 namespace boost {
 namespace urls {

--- a/include/boost/url/url_view.hpp
+++ b/include/boost/url/url_view.hpp
@@ -212,7 +212,7 @@ public:
     url_view(
         String const& s)
         : url_view(
-            string_view(s))
+            detail::to_sv(s))
     {
     }
 

--- a/test/unit/format.cpp
+++ b/test/unit/format.cpp
@@ -15,6 +15,12 @@
 
 #include "test_suite.hpp"
 
+#ifdef BOOST_TEST_CSTR_EQ
+#undef BOOST_TEST_CSTR_EQ
+#define BOOST_TEST_CSTR_EQ(expr1,expr2) \
+    BOOST_TEST_EQ( boost::urls::detail::to_sv(expr1), boost::urls::detail::to_sv(expr2) )
+#endif
+
 namespace boost {
 namespace urls {
 

--- a/test/unit/segments_encoded_ref.cpp
+++ b/test/unit/segments_encoded_ref.cpp
@@ -18,6 +18,12 @@
 
 #include "test_suite.hpp"
 
+#ifdef BOOST_TEST_CSTR_EQ
+#undef BOOST_TEST_CSTR_EQ
+#define BOOST_TEST_CSTR_EQ(expr1,expr2) \
+    BOOST_TEST_EQ( boost::urls::detail::to_sv(expr1), boost::urls::detail::to_sv(expr2) )
+#endif
+
 namespace boost {
 namespace urls {
 

